File: lib/ui/avatar_screen.dart
// -----------------------------------------------------------------------------
//  avatar_screen.dart  (v4 – weitergereicht & validiert)
// -----------------------------------------------------------------------------
//  • AppContext wird durch alle Ebenen gereicht (für Logging, Navigation, Bus)
//  • LayerTabs & ItemTiles haben Zugriff auf globale Konfigurationen
// -----------------------------------------------------------------------------

import 'dart:io';

import 'package:diabetes_kids_app/core/app_context.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:diabetes_kids_app/l10n/gen_l10n/app_localizations.dart';
import '../services/avatar_service.dart';
import '../events/app_events.dart';

class AvatarScreen extends StatelessWidget {
  final AppContext appContext;
  const AvatarScreen({super.key, required this.appContext});

  @override
  Widget build(BuildContext context) => ChangeNotifierProvider.value(
    value: AvatarService.I,
    child: _Body(appContext: appContext),
  );
}

/* ═══════════════════════════════════ UI‑BODY ═══════════════════════════════ */

class _Body extends StatelessWidget {
  final AppContext appContext;
  const _Body({required this.appContext});

  static const _layers = [
    'background',
    'wing',
    'body',
    'head',
    'accessory',
    'weapon'
  ];

  @override
  Widget build(BuildContext context) {
    final l = AppLocalizations.of(context);

    return DefaultTabController(
      length: _layers.length,
      child: Scaffold(
        appBar: AppBar(
          title: Text(l.avatarTitle),
          bottom: TabBar(
            isScrollable: true,
            tabs: _layers.map((e) => Tab(text: _label(l, e))).toList(),
          ),
        ),
        floatingActionButton: FloatingActionButton(
          tooltip: l.avatarRandomize,
          onPressed: () {
            final rnd = _layers
                .map((layer) {
              final candidates = AvatarService.I
                  .itemsForLayer(layer)
                  .where((i) => AvatarService.I.itemUnlocked(i.key))
                  .toList();
              if (candidates.isEmpty) return null;
              candidates.shuffle();
              return candidates.first;
            })
                .whereType<AvatarItem>()
                .toList();

            for (final i in rnd) {
              AvatarService.I.equip(i.layer, i.key);
            }

            // Beispielnutzung AppContext → EventBus oder Logging möglich
            appContext.bus.fire(const AvatarCelebrateEvent());
          },
          child: const Icon(Icons.shuffle),
        ),
        body: TabBarView(
          children: _layers.map((layer) => _LayerTab(layer: layer, appContext: appContext)).toList(),
        ),
      ),
    );
  }

  String _label(AppLocalizations l, String layer) {
    switch (layer) {
      case 'background':
        return l.avatarBackground;
      case 'wing':
        return l.avatarWing;
      case 'body':
        return l.avatarBody;
      case 'head':
        return l.avatarHead;
      case 'accessory':
        return l.avatarAccessory;
      case 'weapon':
        return l.avatarWeapon;
      default:
        return layer;
    }
  }
}

/* ══════════════════════════════════ LAYER TAB ══════════════════════════════ */

class _LayerTab extends StatelessWidget {
  final String layer;
  final AppContext appContext;
  const _LayerTab({required this.layer, required this.appContext});

  @override
  Widget build(BuildContext context) {
    final svc = context.watch<AvatarService>();
    final items = svc.itemsForLayer(layer);

    return Column(
      children: [
        const SizedBox(height: 16),
        const _AvatarPreview(),
        const SizedBox(height: 16),
        Expanded(
          child: GridView.builder(
            padding: const EdgeInsets.all(8),
            gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 4, childAspectRatio: .9),
            itemCount: items.length,
            itemBuilder: (_, i) =>
                _ItemTile(layer: layer, item: items[i], appContext: appContext),
          ),
        ),
      ],
    );
  }
}

/* ══════════════════════════════════ ITEM TILE ══════════════════════════════ */

class _ItemTile extends StatelessWidget {
  final String layer;
  final AvatarItem item;
  final AppContext appContext;
  const _ItemTile({required this.layer, required this.item, required this.appContext});

  @override
  Widget build(BuildContext context) {
    final svc = context.watch<AvatarService>();
    final unlocked = svc.itemUnlocked(item.key);
    final selected = svc.selectedItem(layer) == item.key;

    final imgProvider = item.assetPath.startsWith('/')
        ? FileImage(File(item.assetPath))
        : AssetImage(item.assetPath) as ImageProvider;

    return GestureDetector(
      onTap: unlocked
          ? () {
        svc.equip(layer, item.key);
        appContext.bus.fire(AvatarSpeakEvent(item.name)); // Beispiel
      }
          : null,
      child: Tooltip(
        message: unlocked
            ? item.name
            : _lockedMsg(context, item.unlock, AppLocalizations.of(context)),
        child: Stack(
          alignment: Alignment.center,
          children: [
            AnimatedContainer(
              duration: const Duration(milliseconds: 200),
              margin: const EdgeInsets.all(6),
              decoration: BoxDecoration(
                border: Border.all(
                    color: selected ? Colors.amber : Colors.grey,
                    width: selected ? 3 : 1),
                borderRadius: BorderRadius.circular(8),
              ),
              child: Image(image: imgProvider, fit: BoxFit.contain),
            ),
            if (!unlocked)
              const Icon(Icons.lock, size: 28, color: Colors.redAccent),
          ],
        ),
      ),
    );
  }

  String _lockedMsg(BuildContext ctx, UnlockCondition? c, AppLocalizations l) {
    if (c == null) return l.avatarLocked;
    final pts = c.minPoints;
    final lvl = c.minLevel;
    if (pts != null && lvl != null) {
      return l.avatarLockedPtsLvl(pts, lvl);
    } else if (pts != null) {
      return l.avatarLockedPts(pts);
    } else if (lvl != null) {
      return l.avatarLockedLvl(lvl);
    }
    return l.avatarLocked;
  }
}

/* ══════════════════════════════════ PREVIEW ════════════════════════════════ */

class _AvatarPreview extends StatelessWidget {
  const _AvatarPreview();

  @override
  Widget build(BuildContext context) {
    final svc = context.watch<AvatarService>();
    final imgPaths = svc.getImagePaths();

    return SizedBox(
      width: 200,
      height: 200,
      child: Stack(
        fit: StackFit.expand,
        children: imgPaths.map((p) {
          final provider = p.startsWith('/')
              ? FileImage(File(p))
              : AssetImage(p) as ImageProvider;
          return Image(image: provider);
        }).toList(),
      ),
    );
  }
}

File: lib/AvatarScreen.java
/*
 *  avatar_screen.dart
 *  --------------------------------------------------------------
 *  Vollbild‑Editor:
 *    – Zeigt aktuellen Avatar (zusammengesetzte PNG‑Layer)
 *    – Listen‑Tabs je Layer (Body, Head, Accessory, Weapon, Wing)
 *    – Lock‑Icon für Items, die noch nicht freigeschaltet sind
 *    – Eltern können im Eltern‑Profil neue PNGs hochladen
 *
 *  © 2025 Kids Diabetes Companion – GPL‑3.0‑or‑later
 */

import 'dart:io';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../services/avatar_service.dart';
import '../services/settings_service.dart';

class AvatarScreen extends StatelessWidget {
  const AvatarScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final a = AvatarService.I;

    return ChangeNotifierProvider.value(
      value: _AvatarAdapter(a),
      child: const _AvatarBody(),
    );
  }
}

class _AvatarBody extends StatelessWidget {
  const _AvatarBody({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final a = Provider.of<_AvatarAdapter>(context);
    final layers = ['background', 'wing', 'body', 'head', 'accessory', 'weapon'];

    return DefaultTabController(
      length: layers.length,
      child: Scaffold(
        appBar: AppBar(
          title: const Text('Avatar'),
          bottom: TabBar(
            isScrollable: true,
            tabs: layers.map((l) => Tab(text: l.toUpperCase())).toList(),
          ),
        ),
        body: TabBarView(
          children: layers.map((layer) => _layerTab(layer, a)).toList(),
        ),
      ),
    );
  }

  Widget _layerTab(String layer, _AvatarAdapter a) {
    final items = a.itemsForLayer(layer);
    return Column(
      children: [
        const SizedBox(height: 12),
        _AvatarPreview(a),
        const SizedBox(height: 12),
        Expanded(
          child: GridView.count(
            crossAxisCount: 4,
            children: items.map((it) {
              final unlocked = a.unlocked(it.key);
              final selected = a.isSelected(layer, it.key);
              return GestureDetector(
                onTap: unlocked
                    ? () => a.equip(layer, it.key)
                    : () {},
                child: Stack(
                  alignment: Alignment.center,
                  children: [
                    Container(
                      margin: const EdgeInsets.all(6),
                      decoration: BoxDecoration(
                        border: Border.all(
                          color: selected ? Colors.amber : Colors.grey,
                          width: selected ? 3 : 1,
                        ),
                        borderRadius: BorderRadius.circular(8),
                      ),
                      child: Image(
                        image: it.assetPath.startsWith('/')
                            ? FileImage(File(it.assetPath))
                            : AssetImage(it.assetPath) as ImageProvider,
                      ),
                    ),
                    if (!unlocked)
                      const Icon(Icons.lock, color: Colors.red, size: 24),
                  ],
                ),
              );
            }).toList(),
          ),
        ),
      ],
    );
  }
}

class _AvatarPreview extends StatelessWidget {
  final _AvatarAdapter a;
  const _AvatarPreview(this.a);
  @override
  Widget build(BuildContext context) {
    final layers = ['wing', 'body', 'head', 'accessory', 'weapon'];
    return SizedBox(
      width: 140,
      height: 140,
      child: Stack(
        fit: StackFit.expand,
        children: layers.map((layer) {
          final itemKey = a.selectedItem(layer);
          if (itemKey == null) return const SizedBox.shrink();
          final item = a.byKey(itemKey)!;
          final img = item.assetPath.startsWith('/')
              ? FileImage(File(item.assetPath))
              : AssetImage(item.assetPath) as ImageProvider;
          return Image(image: img);
        }).toList(),
      ),
    );
  }
}

/* ---------------- Adapter ---------------- */

class _AvatarAdapter extends ChangeNotifier {
  final AvatarService _a;
  _AvatarAdapter(this._a);

  List<AvatarItem> itemsForLayer(String layer) =>
      _a.catalog.where((e) => e.layer == layer).toList();

  bool unlocked(String key) => _a.itemUnlocked(key);

  bool isSelected(String layer, String key) =>
      _a.state.equipped[layer] == key;

  String? selectedItem(String layer) => _a.state.equipped[layer];

  AvatarItem? byKey(String key) =>
      _a.catalog.firstWhere((e) => e.key == key, orElse: () => null);

  Future<void> equip(String layer, String key) async {
    await _a.equip(layer, key);
    notifyListeners();
  }
}